pkg/util/scheduler/queue.go

```diff
@@ -0,0 +1,505 @@
+	// tenantQueues stores the queues for each tenant
+	tenantQueues map[string]*tenantQueue
+	// activeTenants is a list of tenants with items in their queues
+	// used for round-robin dequeueing
+	activeTenants *list.List
+	// pendingRequests is a list of dequeue requests waiting for items
+	// used for notifying when items are available
+	pendingRequests *list.List
+	// closed indicates if the queue is closed
+	closed bool
+	// maxSizePerTenant is the maximum number of items per tenant
+	maxSizePerTenant int
+
+	// Metrics
+	queueLength       *prometheus.GaugeVec
+	discardedRequests *prometheus.CounterVec
+}
+
+type firstDequeueItem struct {
+	dequeueRespChan chan dequeueResponse
+	runnable        func()
+	tenantElem      *list.Element
+}
+
+func (s *dispatcherState) shouldExit() bool {
+	return s.closed && s.activeTenants.Len() == 0 && s.pendingRequests.Len() == 0
+}
+
+func (s *dispatcherState) prepareFirstDequeue() firstDequeueItem {
+	item := firstDequeueItem{}
+
+	pendingTenant := s.activeTenants.Front()
+	if !s.closed && pendingTenant != nil && s.pendingRequests.Len() > 0 {
+		tq := pendingTenant.Value.(*tenantQueue)
+		if len(tq.items) > 0 {
+			pendingReq := s.pendingRequests.Front()
+			if pendingReq == nil {
+				return item
+			}
+
+			item.runnable = tq.items[0]
+			item.dequeueRespChan = pendingReq.Value.(*dequeueRequest).respChan
+			item.tenantElem = pendingTenant
+		} else {
+			s.activeTenants.Remove(pendingTenant)
+			tq.activeElement = nil
+		}
+	}
+
+	return item
+}
+
+func (s *dispatcherState) handleEnqueueRequest(req enqueueRequest) {
+	if s.closed {
+		s.discardedRequests.WithLabelValues(req.tenantID, "queue_closed").Inc()
+		req.respChan <- ErrQueueClosed
+		return
+	}
+
+	tq, exists := s.tenantQueues[req.tenantID]
+	if !exists {
+		tq = &tenantQueue{
+			id:    req.tenantID,
+			items: make([]func(), 0, 8),
+		}
+		s.tenantQueues[req.tenantID] = tq
+	}
+
+	if s.maxSizePerTenant > 0 && len(tq.items) >= s.maxSizePerTenant {
+		s.discardedRequests.WithLabelValues(req.tenantID, "queue_full").Inc()
+		req.respChan <- ErrTenantQueueFull
+		return
+	}
+
+	tq.items = append(tq.items, req.runnable)
+	s.queueLength.WithLabelValues(req.tenantID).Set(float64(len(tq.items)))
+
+	if tq.activeElement == nil {
+		tq.activeElement = s.activeTenants.PushBack(tq)
+	}
+
+	req.respChan <- nil
+}
+
+func (s *dispatcherState) handleDequeueRequest(req dequeueRequest) {
+	if s.closed {
+		s.dispatchDequeueResponse(&req, dequeueResponse{ok: false, err: ErrQueueClosed})
+		return
+	}
+
+	if req.ctx.Err() != nil {
+		s.dispatchDequeueResponse(&req, dequeueResponse{ok: false, err: req.ctx.Err()})
+		return
+	}
+
+	s.pendingRequests.PushBack(&req)
+}
+
+func (s *dispatcherState) completeDequeue(tenantElem *list.Element) {
+	reqElem := s.pendingRequests.Front()
+	s.pendingRequests.Remove(reqElem)
+
+	tq := tenantElem.Value.(*tenantQueue)
+	tq.items = tq.items[1:]
+
+	s.queueLength.WithLabelValues(tq.id).Set(float64(len(tq.items)))
```

Comment:
If `len(tq.items) == 0`, shall we remove the series for this tenant, instead of setting it to 0? (I think the correct answer is: set to 0 now, and remove the series after a minute or two, to avoid keeping series forever for idle tenants).